Integrate issue35 with main (#59)
lamps-wg/draft-composite-sigs: draft-ietf-lamps-pq-composite-sigs.md

The key generation functions MUST be executed for both algorithms. Compliant parties MUST NOT use or import component keys that are used in other contexts, combinations, or by themselves (i.e., not only in X.509 certificates).

## Pure Signature Generation {#sec-comp-sig-gen}

Composite schemes' signatures provide important properties for multi-key environments such as non-separability and key-binding. For more information on the additional security properties and their applicability to multi-key or hybrid environments, please refer to {{I-D.hale-pquip-hybrid-signature-spectrums}} and the use of labels as defined in {{Bindel2017}}

> A composite signature's value MUST include two signature components and MUST be in the same order as the components from the corresponding signing key.

Reviewer: ```suggestion
A composite signature's value MUST include two signature components and MUST be in the same order as the components from the corresponding verification public key.
```

The following process is used to generate pure composite signature values and mirrors Algorithm 2 in  [FIPS.204].

~~~
Composite-ML-DSA.Sign (sk, M, ctx) -> (signature)
Explicit Input:
     sk                 Composite private key conisting of signing private keys for each component.

     M                  The Message to be signed, an octet string

     ctx                The Message context string, which defaults to the empty string

Implicit inputs:

    ML-DSA             A placeholder for the specific ML-DSA algorithm and
                       parameter set to use, for example, could be "ML-DSA-65".

    Trad               A placeholder for the specific ML-DSA algorithm and
                       parameter set to use, for example "RSASA-PSS with id-sha256"
                       or "Ed25519".

    Domain             Domain separator value for binding the signature to the Composite OID.
                       See section on Domain Separators below.

Output:
     signature             The composite signature, a CompositeSignatureValue

Signature Generation Process:

   1. If |ctx| > 255:
        return error

>    2. Compute the Message format M' by concatenating the Domain identifier (i.e., the DER encoding of the Composite signature algorithm identifier) with the length of ctx, the value ctx and the M

Reviewer: ```suggestion
   2. Compute the Message M' by concatenating the Domain identifier (i.e., the DER encoding of the Composite signature algorithm identifier) with the length of ctx, the value ctx and the original message M
```

         M' := Domain || len(ctx) || ctx || M

>    3. Separate the private key into componet keys. Note, the exact storage format for composite private keys may be as described in this document, or may be implementation-specific.

Author: typo in component

         (sk1, sk2) := Unmarshal(sk)

   4. Generate the 2 component signatures independently, by calculating the signature over M'
      according to their algorithm specifications that might involve the use of the hash-n-sign paradigm.

         s1 := ML-DSA.Sign( sk1, M', ctx=Domain )
         s2 := Trad.Sign( sk2, M' )

      The Domain is used as the context separator for the ML-DSA.Sign component.

      If either ML-DSA.Sign() or Trad.Sign() return an error, then this process must return an error.

   5. Encode each component signature S1 and S2 into a BIT STRING
      according to its algorithm specification.

          signature := Sequence { s1, s2 }

   6. Output signature

        return signature
~~~
{: # title="Composite-ML-DSA-Sign(sk, M, ctx)"}

It is possible to construct `CompositePrivateKey`(s) to generate signatures from component keys stored in separate software or hardware keystores. Variations in the process to accommodate particular private key storage mechanisms are considered to be conformant to this document so long as it produces the same output as the process sketched above.

### Composite-ML-DSA.Verify {#sec-comp-sig-verify}

> This mode mirrors `ML-DSA.Sign(sk, M, ctx)` defined in Section 5.3 of [FIPS.204]. Verification of a composite signature involves reconstructing the `M'` message by concatenating the composite domain separator "Domain" {{sec-oid-concat}} with the length of the context string `ctx` in bytes, the context string `ctx`, and finally the un-hashed message `M` .

Reviewer: ```suggestion
This mode mirrors `ML-DSA.Verify(pk, M, signature, ctx)` defined in Section 5.3 of [FIPS.204]. Verification of a composite signature involves reconstructing the `M'` message by concatenating the composite domain separator "Domain" {{sec-oid-concat}} with the length of the context string `ctx` in bytes, the context string `ctx`, and finally the original message `M` .
```

                        signature values (S1 and S2) to be verified.
     ctx                The Message context string, which defaults to the empty string

Implicit inputs:

    ML-DSA             A placeholder for the specific ML-DSA algorithm and
                       parameter set to use, for example, could be "ML-DSA-65".

    Trad               A placeholder for the specific ML-DSA algorithm and
                       parameter set to use, for example "RSASA-PSS with id-sha256"
                       or "Ed25519".

    Domain             Domain separator value for binding the signature to the Composite OID.
                       See section on Domain Separators below.

Output:
    Validity (bool)    "Valid signature" (true) if the composite
                        signature is valid, "Invalid signature"
                        (false) otherwise.

Signature Verification Procedure:
   1. Separate the keys and signatures

          (pk1, pk2) := pk
          (s1, s2) := signature

      If Error during Desequencing, or the sequences have
      different numbers of elements, or any of the public keys
      P1 or P2 and the algorithm identifiers A1 or A2 are

>       composite then output "Invalid signature" and stop.

Reviewer: This whole paragraph needs to be re-done. The prohibition of recursive composites is a hold-over from generics and no longer makes sense.

Reviewer: Suggestion:

      If Error during Desequencing, or if any of the component keys or signature values are not of the correct key type or length for the given component algorithm then output "Invalid signature" and stop.

>    2. If |ctx| > 255

Reviewer: The order of steps should be synced with .Sign(). For example, ,Sign() has 

```
1. If |ctx| > 255
2. Compute the Message format M'
3. Separate the private key into componet keys.
```

But here in .Verify(), we have:

```
1. Separate the keys and signatures
2. If |ctx| > 255
3. Format the Message as follows
```

I think the order in .Verify() makes more sense, so we should change .Sign() to match.

Author: I don't agree.   If we want to change them to be the same order, then we change Verify.     In Sign() if ctx > 255 then computing M' would fail because the length of the context can't be encoded into a single byte and computing M' would fail or be incorrect.  Why do all that when a simple check of the length would tell us if we can encode M' properly?    For verify, I don't really think the order matters as separating the keys is independent of the context check, but since the context check is faster, lets just do that one first in verify so they match up...

Author: I made the change to Verify so the two now match.  The first step is now the context check in a cases (both for pure and pre-hash).

        return error

>    3. Format the Message as follows:

Reviewer: ```suggestion
   3. Compute the Message M' by concatenating the Domain identifier (i.e., the DER encoding of the Composite signature algorithm identifier) with the length of ctx, the value ctx and the original message M
```

Signature Generation Process:

   1. If |ctx| > 255:
        return error

   2. Compute the Message format M' by concatenating the Domain identifier (i.e., the DER encoding of the Composite signature algorithm identifier) with the length of the context, the Context, the HashOID and the Hash of the Message.

         M' :=  Domain || len(ctx) || ctx || HashOID || PH(M)

   3. Separate the private key into componet keys. Note, the exact storage format for composite private keys may be as described in this document, or may be implementation-specific.

         (sk1, sk2) := Unmarshal(sk)

   4. Generate the 2 component signatures independently, by calculating the signature over M'
      according to their algorithm specifications that might involve the use of the hash-n-sign paradigm.

         s1 := ML-DSA.Sign( sk1, M', ctx=Domain )
         s2 := Trad.Sign( sk2, M' )

     The Domain is used as the context separator for the ML-DSA.Sign component.

   5. Encode each component signature S1 and S2 into a BIT STRING
      according to its algorithm specification.

          signature := Sequence { s1, s2 }

   6. Output signature

        return signature
~~~

> {: #alg-hash-composite-sign title="HashComposite-ML-DSA-Sign(sk, M, ctx, PH)"}

Reviewer: Again, we should make sure that the order of steps is aligned with the pure algs above.

                       PreHash algorithm (PH) which is passed into the function

Output:
    Validity (bool)    "Valid signature" (true) if the composite
                        signature is valid, "Invalid signature"
                        (false) otherwise.

Signature Verification Procedure::

   1. Separate the keys and signatures

          (pk1, pk2) := pk
          (s1, s2) := signature

      If Error during Desequencing, or the sequences have
      different numbers of elements, or any of the public keys
      P1 or P2 and the algorithm identifiers A1 or A2 are
      composite then output "Invalid signature" and stop.

   2. If |ctx| > 255
        return error

   3. Compute a Hash of the Message

         M' = Domain || len(ctx) || ctx || HashOID || PH(M)

   4. Check each component signature individually, according to its
       algorithm specification.
       If any fail, then the entire signature validation fails.

       if not ML-DSA.Verify( pk1, M', s1, ctx=Domain ) then
            output "Invalid signature"

       if not Trad.Verify( pk2, M', s2 ) then
            output "Invalid signature"

       if all succeeded, then
        output "Valid signature"
~~~

> {: #alg-hash-composite-verify title="Hash-Composite-ML-DSA-Verify(pk, M, signature, ctx, PH)"}

Reviewer: Again, we should make sure that the order of steps is aligned with the pure algs above.

Some use-cases desire the flexibility for clients to use any combination of supported algorithms, while others desire the rigidity of explicitly-specified combinations of algorithms.

> The following tables summarizes the details for each explicit composite signature algorithms:

Reviewer: ```suggestion
The following tables summarize the details for each explicit composite signature algorithms:
```

Therefore &lt;CompSig&gt;.21 is equal to 2.16.840.1.114027.80.8.1.21

> Pure Signature public key types:

Reviewer: ```suggestion
Pure Composite-ML-DSA Signature public key types:
```

| Composite Signature AlgorithmID | OID | First AlgorithmID | Second AlgorithmID | Second Alg PreHash |
| ----------- | ----------- | ----------- |  ----------- | ----------- |
| id-MLDSA44-RSA2048-PSS-SHA256      | &lt;CompSig&gt;.21 | id-ML-DSA-44  | id-RSASA-PSS with id-sha256 | id-sha256 |
| id-MLDSA44-RSA2048-PKCS15-SHA256    | &lt;CompSig&gt;.22 | id-ML-DSA-44  | sha256WithRSAEncryption | id-sha256 |

> | id-MLDSA44-Ed25519                  | &lt;CompSig&gt;.23 | id-ML-DSA-44  | id-Ed25519 | Not Needed |

Reviewer: ```suggestion
| id-MLDSA44-Ed25519                  | &lt;CompSig&gt;.23 | id-ML-DSA-44  | id-Ed25519 | None |
```

Author: Group agreed with this change.

| id-MLDSA44-ECDSA-P256-SHA256         | &lt;CompSig&gt;.24 | id-ML-DSA-44  | ecdsa-with-SHA256 with secp256r1 | id-sha256 |
| id-MLDSA65-RSA3072-PSS-SHA512           | &lt;CompSig&gt;.26 | id-ML-DSA-65 | id-RSASA-PSS with id-sha512 | id-sha512 |
| id-MLDSA65-RSA3072-PKCS15-SHA512        | &lt;CompSig&gt;.27  | id-ML-DSA-65 | sha512WithRSAEncryption | id-sha512 |
| id-MLDSA65-RSA4096-PSS-SHA512           | &lt;CompSig&gt;.34 | id-ML-DSA-65 | id-RSASA-PSS with id-sha512 | id-sha512 |
| id-MLDSA65-RSA4096-PKCS15-SHA512        | &lt;CompSig&gt;.35  | id-ML-DSA-65 | sha512WithRSAEncryption | id-sha512 |
| id-MLDSA65-ECDSA-P384-SHA512            | &lt;CompSig&gt;.28  | id-ML-DSA-65 | ecdsa-with-SHA512 with secp384r1 | id-sha512 |
| id-MLDSA65-ECDSA-brainpoolP256r1-SHA512 | &lt;CompSig&gt;.29  | id-ML-DSA-65 | ecdsa-with-SHA512 with brainpoolP256r1 | id-sha512 |

> | id-MLDSA65-Ed25519                      | &lt;CompSig&gt;.30  | id-ML-DSA-65 | id-Ed25519 | Not Needed |

Reviewer: ```suggestion
| id-MLDSA65-Ed25519                      | &lt;CompSig&gt;.30  | id-ML-DSA-65 | id-Ed25519 | None |
```

| id-MLDSA87-ECDSA-P384-SHA512            | &lt;CompSig&gt;.31  | id-ML-DSA-87 | ecdsa-with-SHA512 with secp384r1 | id-sha512|
| id-MLDSA87-ECDSA-brainpoolP384r1-SHA512 | &lt;CompSig&gt;.32 | id-ML-DSA-87 | ecdsa-with-SHA512 with brainpoolP384r1 | id-sha512 |

> | id-MLDSA87-Ed448                        | &lt;CompSig&gt;.33 | id-ML-DSA-87 | id-Ed448 | Not Needed |

Reviewer: ```suggestion
| id-MLDSA87-Ed448                        | &lt;CompSig&gt;.33 | id-ML-DSA-87 | id-Ed448 | None |
```

{: #tab-sig-algs title="Pure ML-DSA Composite Signature Algorithms"}

The table above contains everything needed to implement the listed pure ML-DSA composite signature algorithms.  The hash value indicated is used only by the Second algorithm if needed.  See the ASN.1 module in section {{sec-asn1-module}} for the explicit definitions of the above Composite signature algorithms.

> HashML-DSA Signature public key types:

Reviewer: ```suggestion
HashComposite-ML-DSA Signature public key types:
```
